src/plugins/i18n/server/i18n.js

```diff
@@ -0,0 +1,144 @@
+    var translationFile = translationStorePluginPath + '/' + translationFileName;
+    fs.readFile(translationFile, function (err, translationStr) {
+      if (err) return callback(err);
+      var translationJson = [];
+      try {
+        translationJson = JSON.parse(translationStr);
+      } catch (e) {
+        return callback('Bad ' + language + ' translation strings for plugin ' + pluginName + '. Error: ' + err);
+      }
+      return callback(null, translationJson);
+    });
+  },
+
+  getPluginTranslationDetails: function (pluginTranslationPath, translationFiles, languageList, callback) {
+    try {
+      getFilesRecursivelyFromTopDir(pluginTranslationPath, translationFiles, languageList);
+    } catch (err) {
+      return callback(err);
+    }
+    return callback(null);
+  },
+
+  getPluginTranslationStoragePath: function (pluginName) {
+    return TRANSLATION_STORE_PATH + '/' + pluginName;
+  }
+
+};
+
+function saveTranslationToFile(translationFullFileName, translationJson) {
+  var jsonToWrite = [];
+  if (fs.existsSync(translationFullFileName)) {
+    var prevTranslationJson = require(translationFullFileName);
+    jsonToWrite = prevTranslationJson.concat(translationJson);
+  } else {
+    jsonToWrite = translationJson;
+  }
+  fs.writeFileSync(translationFullFileName, JSON.stringify(jsonToWrite, null, 4));
+}
+
+function getFilesRecursivelyFromTopDir(topDir, translationFiles, languageList) {
+  fs.readdirSync(topDir).forEach(function (name) {
+    var fullPath = path.join(topDir, name);
+    var stat = fs.statSync(fullPath);
+    if (stat.isDirectory()) {
+      getTranslationDetailsFromDirectory(fullPath, translationFiles, languageList);
+    }
+  });
+}
+
+function getTranslationDetailsFromDirectory(fullPath, translationFiles, languageList) {
+  var files = getFilesFromDir(fullPath);
+  var fileLength = files.length;
+  for (var i = 0; i < fileLength; i++) {
+    var fullFilePath = files[i];
+    var fileName = getFileName(fullFilePath);
+    var fileExt = fileName.split('.').pop();
+    if (fileName === fileExt) continue;
+    if (fileExt !== TRANSLATION_FILE_EXTENSION) continue;
+    translationFiles.push(fullFilePath);
+    var lang  = fileName.substr(0, fileName.lastIndexOf('.'));
+    if (languageList.indexOf(lang) !== -1) {
+      continue;
+    }
+    languageList.push(lang);
+  }
+}
+
+function getFilesFromDir(dir) {
+  var fileList = [];
+
+  var files = fs.readdirSync(dir);
+  for (var i in files) {
+    if (!files.hasOwnProperty(i)) continue;
+    var name = dir + '/' + files[i];
+    if (!fs.statSync(name).isDirectory()) {
+      fileList.push(name);
+    }
+  }
+  return fileList;
+}
+
+function getFileName(fullPath) {
+  return fullPath.replace(/^.*[\\\/]/, '');
+}
+
+// Added this function because 'mkdirp' does not add more than 2 subdirectories
+function createDirectoriesRecursively(fullDir) {
+  process.execSync('mkdir -p ' + fullDir);
```

Comment:
@srl295, good point, thanks. I was having issue previously but may have been that I was using the asynchronous version of 'mkdirp' in a synchronous fashion. I have updated the code now to use 'mkdirp.sync' and is working as expected.